components/hab/src/main.rs

```diff
@@ -841,6 +843,51 @@ fn sub_pkg_sign(ui: &mut UI, m: &ArgMatches<'_>) -> Result<()> {
     command::pkg::sign::start(ui, &pair, &src, &dst)
 }
 
+fn sub_pkg_bulkupload(ui: &mut UI, m: &ArgMatches<'_>) -> Result<()> {
+    let upload_dir = bulkupload_dir_from_matches(m);
+    // upload_dir is required via clap and exists
+    let artifact_path = upload_dir.as_ref().unwrap().join("artifacts");
+    let key_path = upload_dir.as_ref().unwrap().join("keys");
+    let url = bldr_url_from_matches(&m)?;
+    let additional_release_channel = channel_from_matches(&m);
+    let force_upload = m.is_present("FORCE");
+    let auto_build = if m.is_present("AUTO_BUILD") {
+        BuildOnUpload::PackageDefault
+    } else {
+        BuildOnUpload::Disable
+    };
+    let token = auth_token_param_or_env(&m)?;
+    let options = glob::MatchOptions { case_sensitive:              true,
+                                       require_literal_separator:   true,
+                                       require_literal_leading_dot: true, };
+    let artifact_paths =
+        vec_from_glob_with(&artifact_path.join("*.hart").display().to_string(), options);
+
+    ui.begin(format!("Preparing to upload artifacts to the '{}' channel on {}",
+                     additional_release_channel.clone()
+                                               .unwrap_or_else(ChannelIdent::unstable),
+                     url))?;
+    ui.status(Status::Using,
+              format!("{} for artifacts and {} for signing keys.",
+                      &artifact_path.display(),
+                      &key_path.display()))?;
+    ui.status(Status::Found,
+              format!("{} artifacts for upload.", artifact_paths.len()))?;
+
+    for artifact_path in &artifact_paths {
+        command::pkg::upload::start(ui,
+                                    &url,
+                                    &additional_release_channel,
+                                    &token,
+                                    &artifact_path,
+                                    force_upload,
+                                    auto_build,
+                                    &key_path)?;
```

Comment:
It's unnecessary to have the `&` both before `artifact_paths` (in L877) *and* before `artifact_path`.

The `for` loop implicitly calls [`into_iter`](http://doc.rust-lang.org/1.38.0/std/iter/trait.IntoIterator.html) on the value after `in` which will return an iterator over `PathBuf`s given `artifact_paths` and an iterator over `&PathBuf`s given `&artifact_paths`. There's a good explanation here: http://xion.io/post/code/rust-for-loop.html.

I suggest removing the `&` from  `artifact_paths` and leaving it on `&artifact_path`.

Reply:
Addressed in https://github.com/habitat-sh/habitat/pull/6958/commits/bc5b9642d03ce480fd4573ae05eaef724b3c12a7